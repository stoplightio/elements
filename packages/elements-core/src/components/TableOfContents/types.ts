export type {
  CustomLinkComponent,
  TableOfContentsDivider,
  TableOfContentsExternalLink,
  TableOfContentsGroup,
  TableOfContentsGroupItem,
  TableOfContentsItem,
  TableOfContentsNode,
  TableOfContentsNodeGroup,
  TableOfContentsProps,
} from '@stoplight/elements-utils';
