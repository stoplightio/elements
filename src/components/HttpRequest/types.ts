export type ParamType = 'query' | 'path' | 'header';
