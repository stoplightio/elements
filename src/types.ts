import { JsonPath, NodeType } from '@stoplight/types';
import { IconName } from '@stoplight/ui-kit';
import { IContentsNode } from '@stoplight/ui-kit/TableOfContents/types';
import { Edge, Node } from 'vis';
export { IComponentMapping } from '@stoplight/markdown-viewer';

export interface IDeserializedSrn {
  service: string;
  org: string;
  project: string;
  uri: string;
}

export interface IChange {
  createdAt: string;
  semver: string;
  message: string;
}

export interface IDiagnostic {
  location: { uri: string; jsonPath?: JsonPath };
  message: string;
}

export interface INodeInfo<D = unknown> {
  id: number;
  type: NodeType;
  name: string;
  srn: string;
  data: D;

  errors?: IDiagnostic[];
  changes?: IChange[];
  version?: string;
  versions?: string[];
  tags?: string[];
}

export interface IProjectNode {
  type: NodeType;
  srn: string;
  name: string;
  id: number | string;

  latestVersion?: string;
  versions?: string[];
  tags?: string[];
  data?: string;
  summary?: string;
}

export interface IContentsNodeWithId extends IContentsNode {
  id: number | string;
}

export type ProjectNodeWithUri = IProjectNode & { uri: string };

export type IconMapType = NodeType | 'group' | 'divider' | 'item';
export type NodeIconMapping = { [type in IconMapType]?: IconName };

export type DocsNodeType = NodeType | 'json_schema' | 'http';

export interface IPageHeading {
  id: string;
  title: string;
  depth: number;
}

export interface IPaginatedResponse<T> {
  items: T[];
  pageInfo: {
    hasNextPage: boolean;
    hasPreviousPage: boolean;
    startCursor: string;
    endCursor: string;
  };
  totalCount: number;
}

export interface IVisGraph {
  nodes: Node[];
  edges: Edge[];
}

export interface INodeGraph {
  nodes: IGraphNode[];
  edges: IGraphEdge[];
}

export interface IGraphNode {
  groupNodeId: number;
  name: string;
  srn: string;
  uri: string;
  depth: number;
  type: NodeType | string;
  version: string;

  projectName: string;
  groupSlug: string;
}

export interface IGraphEdge {
  fromGroupNodeId: number;
  fromPath: string;
  toGroupNodeId: number;
  toPath: string;
}
