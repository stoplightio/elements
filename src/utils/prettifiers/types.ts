export type Prettifier<T> = (content: T) => string | Promise<string>;
