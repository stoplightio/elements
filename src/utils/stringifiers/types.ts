export type Stringifier<T> = (content: T) => string;
