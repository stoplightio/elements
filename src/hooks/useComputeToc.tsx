import { Dictionary, NodeType } from '@stoplight/types';
import { IContentsNode } from '@stoplight/ui-kit/TableOfContents/types';
import { compact, escapeRegExp, sortBy, startCase, words } from 'lodash';
import * as React from 'react';
import { IconsContext } from '../containers/Provider';
import { IContentsNodeWithId, IProjectNode, ProjectNodeWithUri } from '../types';
import { NodeIconMapping } from '../types';
import { deserializeSrn } from '../utils/srns';

const README_REGEXP = new RegExp(`${escapeRegExp('README.md')}$`, 'i'); // Regex to get the README file

/**
 * Memoized hook that computes a tree structure from an array of nodes
 */
export function useComputeToc(nodes: IProjectNode[]) {
  const icons = React.useContext(IconsContext);
  return React.useMemo(() => computeToc(nodes, icons), [nodes]);
}

/**
 * Sorts project nodes into a flat array
 */

export function computeToc(_nodes: IProjectNode[], icons: NodeIconMapping): IContentsNodeWithId[] {
  // There is a chance that we pass an empty array
  if (!_nodes.length) return [];

  // Add uri to each node since it's used heavily in this function
  const nodes: ProjectNodeWithUri[] = _nodes.map(n => ({ ...n, uri: deserializeSrn(n.srn).uri }));

  let contents: IContentsNodeWithId[] = [];
  const folders: string[] = [];
  const rootNodes: IContentsNodeWithId[] = []; // These nodes will appear at the top of the tree

  // Grab the root level README and put it at the top of the folder
  const readmeNode = nodes.find(node => {
    return README_REGEXP.test(node.uri) && compact(node.uri.split('/')).length === 1;
  });
  if (readmeNode) {
    rootNodes.push({
      id: readmeNode.id,
      name: readmeNode.name,
      depth: 0,
      type: 'item',
      icon: icons[readmeNode.type] || icons.item,
      href: readmeNode.srn,
    });
  }

  /** Docs folder */
  const docsNodes = sortBy(
    nodes.filter(node => /^\/docs/.test(node.uri) && node.type === NodeType.Article),
    'srn',
  );
  for (const nodeIndex in docsNodes) {
    if (!docsNodes[nodeIndex]) continue;
    const node = docsNodes[nodeIndex];

    // Strip off the /docs since we ignore that folder
    const uri = node.uri.replace(/^\/docs\//, '');
    const parts = uri.split('/');

    // Handle adding the parent folders if we haven't already added them
    if (parts.length > 1) {
      // All the path parts not including the file name
      const pathToItem = parts.slice(0, -1);
      for (const pathIndex in pathToItem) {
        if (!pathToItem[pathIndex]) continue;

        // Create a folder if one doesn't already exist
        const folderName = pathToItem[pathIndex];
        if (!folders.includes(`${folderName}/${pathIndex}`)) {
          folders.push(`${folderName}/${pathIndex}`);
          contents.push({
            id: `${nodeIndex}-${pathIndex}`,
            name: startCase(words(folderName).join(' ')),
            depth: Number(pathIndex),
            type: 'group',
            icon: icons.group,
          });
        }
      }

      contents.push({
        id: node.id,
        name: node.name,
        depth: parts.length - 1,
        type: 'item',
        icon: icons[node.type] || icons.item,
        href: node.srn,
      });
    } else {
      // if our node only has one part, it must not be listed in a folder! Lets add it to a group that we will push onto the front of the stack at the end of this loop
      rootNodes.push({
        id: node.id,
        name: node.name,
        depth: 0,
        type: 'item',
        icon: icons[node.type] || icons.item,
        href: node.srn,
      });
    }
  }

  // Add the root nodes to the top of the tree
  contents = rootNodes.concat(contents);

  /** Reference folder */
  const httpServiceNodes = sortBy(
    nodes.filter(n => n.type === NodeType.HttpService),
    'name',
  );
  for (const httpServiceNode of httpServiceNodes) {
    const parentUriRegexp = new RegExp(`^${escapeRegExp(httpServiceNode.uri)}\/`, 'i');

    const childNodes = nodes.filter(node => parentUriRegexp.test(node.uri) && node.type !== NodeType.HttpService);
    if (!childNodes.length) continue;

    contents.push({
      id: httpServiceNode.id,
      name: httpServiceNode.name,
      depth: 0,
      type: 'divider',
      icon: icons[httpServiceNode.type] || icons.divider,
    });
    contents.push({
      id: `${httpServiceNode.id}-overview`,
      name: 'Overview',
      depth: 0,
      icon: icons.item,
      type: 'item',
      href: httpServiceNode.srn,
    });

    const tags: Dictionary<IProjectNode[], string> = {};
    const other = [];

    /** Group by Tags */
    for (const childNode of childNodes) {
      if (childNode.tags && childNode.tags.length) {
        const tag = childNode.tags[0].toLowerCase();
        if (!tags[tag]) {
          tags[tag] = [];
        }

        tags[tag].push(childNode);
      } else {
        other.push(childNode);
      }
    }

    /** Add tag groups to the tree */
    const sortedTags = sortBy(Object.keys(tags));
    for (const tagIndex in sortedTags) {
      if (!sortedTags[tagIndex]) continue;
      const tag = sortedTags[tagIndex];

      contents.push({
        id: `${httpServiceNode.id}-${tag}-${tagIndex}`,
        name: startCase(tag),
        depth: 0,
        type: 'group',
        icon: icons.group,
      });

      for (const tagChild of tags[tag]) {
        contents.push({
          id: tagChild.id,
          name: tagChild.name,
          depth: 1,
          icon: icons[tagChild.type] || icons.item,
          type: 'item',
          href: tagChild.srn,
        });
      }
    }

    /** Group whatever is left into "Other" */
    if (other.length) {
      contents.push({
        id: `${httpServiceNode.id}-other`,
        name: 'Other',
        depth: 0,
        type: 'group',
        icon: icons.group,
      });

      for (const otherChild of other) {
        contents.push({
          id: otherChild.id,
          name: otherChild.name,
          depth: 1,
          icon: icons[otherChild.type] || icons.item,
          type: 'item',
          href: otherChild.srn,
        });
      }
    }
  }

  /** Models folder */
  const modelContents: IContentsNodeWithId[] = [];
  const modelNodes = sortBy(
    nodes.filter(n => n.type === NodeType.Model),
    'name',
  );
  for (const modelNode of modelNodes) {
    // Only add models that aren't already in the tree
    if (contents.find(n => n.href === modelNode.srn)) continue;

    modelContents.push({
      id: modelNode.id,
      name: modelNode.name,
      href: modelNode.srn,
      depth: 0,
      type: 'item',
      icon: icons[modelNode.type] || icons.item,
    });
  }

  if (modelContents.length) {
    contents.push({
      id: 'models',
      name: 'Models',
      depth: 0,
      type: 'divider',
    });

    contents = contents.concat(modelContents);
  }

  return contents;
}
