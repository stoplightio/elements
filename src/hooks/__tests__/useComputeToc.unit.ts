import { NodeType } from '@stoplight/types';
import * as fs from 'fs';
import * as path from 'path';
import { computeToc } from '../useComputeToc';

const fixturesPath = '../../__fixtures__/table-of-contents';
const files = fs.readdirSync(path.resolve(__dirname, fixturesPath));

describe('computeToc', () => {
  files.forEach(filePath => {
    describe(filePath, () => {
      const file = require(`${fixturesPath}/${filePath}`);

      it('should return the correct table of contents', () => {
        expect(
          computeToc(file.nodes, {
            group: 'folder-close',
            divider: 'chevron-right',
            item: 'document',
          }),
        ).toMatchSnapshot();
      });

      it('should return the correct icons based on node type', () => {
        expect(
          computeToc(file.nodes, {
            article: 'cube',
            model: 'box',
            http_operation: 'download',
            http_service: 'badge',
            http_server: 'cloud',
          }),
        ).toMatchSnapshot();
      });
    });
  });
});

describe('alphabetized computeToc', () => {
  it('should alphabetically sort tagged models', () => {
    expect(
      computeToc(
        [
          {
            id: 11,
            type: NodeType.HttpService,
            name: 'Another API',
            srn: 'sl/org/project/openapi.v1.yml',
            latestVersion: '1.0',
          },
          {
            id: 12,
            type: NodeType.HttpOperation,
            name: 'List Things',
            srn: 'sl/org/project/openapi.v1.yml/paths/~1things/get',
            tags: ['things'],
          },
          {
            id: 13,
            type: NodeType.Model,
            name: 'Thing',
            srn: 'sl/org/project/openapi.v1.yml/definitions/thing',
            tags: ['things'],
          },
          {
            id: 13,
            type: NodeType.Model,
            name: 'Another Thing',
            srn: 'sl/org/project/openapi.v1.yml/definitions/another-thing',
            tags: ['things'],
          },
        ],
        {
          group: 'folder-close',
          divider: 'chevron-right',
          item: 'document',
        },
      ),
    ).toMatchSnapshot();
  });
});
