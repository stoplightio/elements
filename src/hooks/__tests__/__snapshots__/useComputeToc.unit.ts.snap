// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alphabetized computeToc should alphabetically sort tagged models 1`] = `
Array [
  Object {
    "depth": 0,
    "icon": "chevron-right",
    "id": 11,
    "meta": "v1.0",
    "name": "Another API",
    "type": "divider",
  },
  Object {
    "depth": 0,
    "href": "sl/org/project/openapi.v1.yml",
    "icon": "document",
    "id": "11-overview",
    "name": "Overview",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": "folder-close",
    "id": "11-things-0",
    "name": "Things",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/openapi.v1.yml/definitions/another-thing",
    "icon": "document",
    "id": 13,
    "name": "Another Thing",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/openapi.v1.yml/paths/~1things/get",
    "icon": "document",
    "id": 12,
    "name": "List Things",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/openapi.v1.yml/definitions/thing",
    "icon": "document",
    "id": 13,
    "name": "Thing",
    "type": "item",
  },
]
`;

exports[`computeToc nested.ts should return the correct icons based on node type 1`] = `
Array [
  Object {
    "depth": 0,
    "icon": undefined,
    "id": "0-0",
    "name": "One Level Deep",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/one-level-deep/file-1.md",
    "icon": "cube",
    "id": "gzv028oc",
    "name": "One level Deep - 1",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/one-level-deep/file-2.md",
    "icon": "cube",
    "id": "gzv028oc",
    "name": "One level Deep - 2",
    "type": "item",
  },
  Object {
    "depth": 1,
    "icon": undefined,
    "id": "2-1",
    "name": "Two Levels Deep",
    "type": "group",
  },
  Object {
    "depth": 2,
    "href": "gh/stoplightio/studio/docs/one-level-deep/two-levels-deep/file-1.md",
    "icon": "cube",
    "id": "gzv028oc",
    "name": "Two levels Deep - 1",
    "type": "item",
  },
  Object {
    "depth": 2,
    "href": "gh/stoplightio/studio/docs/one-level-deep/two-levels-deep/file-2.md",
    "icon": "cube",
    "id": "gzv028oc",
    "name": "Two levels Deep - 2",
    "type": "item",
  },
  Object {
    "depth": 2,
    "icon": undefined,
    "id": "4-2",
    "name": "Three Levels Deep",
    "type": "group",
  },
  Object {
    "depth": 3,
    "icon": undefined,
    "id": "4-3",
    "name": "Four Levels Deep",
    "type": "group",
  },
  Object {
    "depth": 4,
    "href": "gh/stoplightio/studio/docs/one-level-deep/two-levels-deep/three-levels-deep/four-levels-deep/file-1.md",
    "icon": "cube",
    "id": "gzv028oc",
    "name": "Four levels Deep - 1",
    "type": "item",
  },
  Object {
    "depth": 4,
    "href": "gh/stoplightio/studio/docs/one-level-deep/two-levels-deep/three-levels-deep/four-levels-deep/file-2.md",
    "icon": "cube",
    "id": "gzv028oc",
    "name": "Four levels Deep - 2",
    "type": "item",
  },
]
`;

exports[`computeToc nested.ts should return the correct table of contents 1`] = `
Array [
  Object {
    "depth": 0,
    "icon": "folder-close",
    "id": "0-0",
    "name": "One Level Deep",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/one-level-deep/file-1.md",
    "icon": "document",
    "id": "gzv028oc",
    "name": "One level Deep - 1",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/one-level-deep/file-2.md",
    "icon": "document",
    "id": "gzv028oc",
    "name": "One level Deep - 2",
    "type": "item",
  },
  Object {
    "depth": 1,
    "icon": "folder-close",
    "id": "2-1",
    "name": "Two Levels Deep",
    "type": "group",
  },
  Object {
    "depth": 2,
    "href": "gh/stoplightio/studio/docs/one-level-deep/two-levels-deep/file-1.md",
    "icon": "document",
    "id": "gzv028oc",
    "name": "Two levels Deep - 1",
    "type": "item",
  },
  Object {
    "depth": 2,
    "href": "gh/stoplightio/studio/docs/one-level-deep/two-levels-deep/file-2.md",
    "icon": "document",
    "id": "gzv028oc",
    "name": "Two levels Deep - 2",
    "type": "item",
  },
  Object {
    "depth": 2,
    "icon": "folder-close",
    "id": "4-2",
    "name": "Three Levels Deep",
    "type": "group",
  },
  Object {
    "depth": 3,
    "icon": "folder-close",
    "id": "4-3",
    "name": "Four Levels Deep",
    "type": "group",
  },
  Object {
    "depth": 4,
    "href": "gh/stoplightio/studio/docs/one-level-deep/two-levels-deep/three-levels-deep/four-levels-deep/file-1.md",
    "icon": "document",
    "id": "gzv028oc",
    "name": "Four levels Deep - 1",
    "type": "item",
  },
  Object {
    "depth": 4,
    "href": "gh/stoplightio/studio/docs/one-level-deep/two-levels-deep/three-levels-deep/four-levels-deep/file-2.md",
    "icon": "document",
    "id": "gzv028oc",
    "name": "Four levels Deep - 2",
    "type": "item",
  },
]
`;

exports[`computeToc studio.ts should return the correct icons based on node type 1`] = `
Array [
  Object {
    "depth": 0,
    "href": "gh/stoplightio/studio/README.md",
    "icon": "cube",
    "id": "readme",
    "name": "Studio Docs",
    "type": "item",
  },
  Object {
    "depth": 0,
    "href": "gh/stoplightio/studio/docs/ui-overview.md",
    "icon": "cube",
    "id": "gzv028oc",
    "name": "UI Overview",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": undefined,
    "id": "0-0",
    "name": "Designing Apis",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/designing-apis/10-getting-started.md",
    "icon": "cube",
    "id": "ks8cwyvs",
    "name": "Modeling Introduction",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/designing-apis/directory-structure.md",
    "icon": "cube",
    "id": "vxeympcn",
    "name": "Project Structure (Design & Modeling)",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/designing-apis/http-endpoints.md",
    "icon": "cube",
    "id": "bb4i9mek",
    "name": "Working with Models",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/designing-apis/using-references.md",
    "icon": "cube",
    "id": "ovxlx772",
    "name": "Using References",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/designing-apis/validation-style-guide.md",
    "icon": "cube",
    "id": "c2riagde",
    "name": "Validation and Linting",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": undefined,
    "id": "6-0",
    "name": "Workflow",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/workflow/customize-interface.md",
    "icon": "cube",
    "id": "yok9mcyg",
    "name": "Customize Interface Panels",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/workflow/working-with-existing-project.md",
    "icon": "cube",
    "id": "kfc8upgh",
    "name": "Working with an Existing Project",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/workflow/working-with-files.md",
    "icon": "cube",
    "id": "3rxli9io",
    "name": "Working with Files",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/workflow/working-with-git.md",
    "icon": "cube",
    "id": "rb6z0v9w",
    "name": "Working with Git",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": undefined,
    "id": "10-0",
    "name": "Writing Tech Docs",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/writing-tech-docs/10-getting-started.md",
    "icon": "cube",
    "id": "lnciknzw",
    "name": "Documentation Quickstart Guide",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/writing-tech-docs/directory-structure.md",
    "icon": "cube",
    "id": "t6694f4v",
    "name": "Directory Structure (Documentation)",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/writing-tech-docs/linking-between-articles.md",
    "icon": "cube",
    "id": "4znkyjxn",
    "name": "Linking Between Articles",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/writing-tech-docs/publishing.md",
    "icon": "cube",
    "id": "20p2qlve",
    "name": "Publishing in Studio",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/writing-tech-docs/stoplight-flavored-markdown.md",
    "icon": "cube",
    "id": "cu1rik0j",
    "name": "Using Markdown in Documentation",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/writing-tech-docs/using-images.md",
    "icon": "cube",
    "id": "stz6uwdt",
    "name": "Using Images",
    "type": "item",
  },
]
`;

exports[`computeToc studio.ts should return the correct table of contents 1`] = `
Array [
  Object {
    "depth": 0,
    "href": "gh/stoplightio/studio/README.md",
    "icon": "document",
    "id": "readme",
    "name": "Studio Docs",
    "type": "item",
  },
  Object {
    "depth": 0,
    "href": "gh/stoplightio/studio/docs/ui-overview.md",
    "icon": "document",
    "id": "gzv028oc",
    "name": "UI Overview",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": "folder-close",
    "id": "0-0",
    "name": "Designing Apis",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/designing-apis/10-getting-started.md",
    "icon": "document",
    "id": "ks8cwyvs",
    "name": "Modeling Introduction",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/designing-apis/directory-structure.md",
    "icon": "document",
    "id": "vxeympcn",
    "name": "Project Structure (Design & Modeling)",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/designing-apis/http-endpoints.md",
    "icon": "document",
    "id": "bb4i9mek",
    "name": "Working with Models",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/designing-apis/using-references.md",
    "icon": "document",
    "id": "ovxlx772",
    "name": "Using References",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/designing-apis/validation-style-guide.md",
    "icon": "document",
    "id": "c2riagde",
    "name": "Validation and Linting",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": "folder-close",
    "id": "6-0",
    "name": "Workflow",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/workflow/customize-interface.md",
    "icon": "document",
    "id": "yok9mcyg",
    "name": "Customize Interface Panels",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/workflow/working-with-existing-project.md",
    "icon": "document",
    "id": "kfc8upgh",
    "name": "Working with an Existing Project",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/workflow/working-with-files.md",
    "icon": "document",
    "id": "3rxli9io",
    "name": "Working with Files",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/workflow/working-with-git.md",
    "icon": "document",
    "id": "rb6z0v9w",
    "name": "Working with Git",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": "folder-close",
    "id": "10-0",
    "name": "Writing Tech Docs",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/writing-tech-docs/10-getting-started.md",
    "icon": "document",
    "id": "lnciknzw",
    "name": "Documentation Quickstart Guide",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/writing-tech-docs/directory-structure.md",
    "icon": "document",
    "id": "t6694f4v",
    "name": "Directory Structure (Documentation)",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/writing-tech-docs/linking-between-articles.md",
    "icon": "document",
    "id": "4znkyjxn",
    "name": "Linking Between Articles",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/writing-tech-docs/publishing.md",
    "icon": "document",
    "id": "20p2qlve",
    "name": "Publishing in Studio",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/writing-tech-docs/stoplight-flavored-markdown.md",
    "icon": "document",
    "id": "cu1rik0j",
    "name": "Using Markdown in Documentation",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "gh/stoplightio/studio/docs/writing-tech-docs/using-images.md",
    "icon": "document",
    "id": "stz6uwdt",
    "name": "Using Images",
    "type": "item",
  },
]
`;

exports[`computeToc studio-template.ts should return the correct icons based on node type 1`] = `
Array [
  Object {
    "depth": 0,
    "href": "sl/org/project/docs/getting-started.md",
    "icon": "cube",
    "id": 1,
    "name": "Getting Started",
    "type": "item",
  },
  Object {
    "depth": 0,
    "href": "sl/org/project/docs/welcome.md",
    "icon": "cube",
    "id": 0,
    "name": "Welcome",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": undefined,
    "id": "1-0",
    "name": "Guides",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/docs/guides/authentication.md",
    "icon": "cube",
    "id": 2,
    "name": "Authentication",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": undefined,
    "id": "2-0",
    "name": "Tutorials",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/docs/tutorials/billing.md",
    "icon": "cube",
    "id": 4,
    "name": "Billing",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/docs/tutorials/signup.md",
    "icon": "cube",
    "id": 3,
    "name": "Signup",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": "badge",
    "id": 11,
    "meta": "v1.0",
    "name": "Another API",
    "type": "divider",
  },
  Object {
    "depth": 0,
    "href": "sl/org/project/openapi.v1.yml",
    "icon": undefined,
    "id": "11-overview",
    "name": "Overview",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": undefined,
    "id": "11-things-0",
    "name": "Things",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/openapi.v1.yml/paths/~1things/get",
    "icon": "download",
    "id": 12,
    "name": "List Things",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/openapi.v1.yml/definitions/thing",
    "icon": "box",
    "id": 13,
    "name": "Thing",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": "badge",
    "id": 5,
    "name": "Todos API",
    "type": "divider",
  },
  Object {
    "depth": 0,
    "href": "sl/org/project/reference/todos/openapi.v1.yml",
    "icon": undefined,
    "id": "5-overview",
    "name": "Overview",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": undefined,
    "id": "5-users-0",
    "name": "Users",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/reference/todos/openapi.v1.yml/paths/~1users/get",
    "icon": "download",
    "id": 6,
    "name": "Get User",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": undefined,
    "id": "5-other",
    "name": "Other",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/reference/todos/openapi.v1.yml/paths/~1todos/get",
    "icon": "download",
    "id": 8,
    "name": "Get Todo",
    "type": "item",
  },
  Object {
    "depth": 0,
    "id": "models",
    "name": "Models",
    "type": "divider",
  },
  Object {
    "depth": 0,
    "href": "sl/org/project/models/example",
    "icon": "box",
    "id": 10,
    "meta": "v1.0",
    "name": "Example",
    "type": "item",
  },
  Object {
    "depth": 0,
    "href": "sl/org/project/reference/todos/models/todos",
    "icon": "box",
    "id": 9,
    "name": "Todo",
    "type": "item",
  },
  Object {
    "depth": 0,
    "href": "sl/org/project/reference/todos/models/user",
    "icon": "box",
    "id": 7,
    "meta": "v1.0",
    "name": "User",
    "type": "item",
  },
]
`;

exports[`computeToc studio-template.ts should return the correct table of contents 1`] = `
Array [
  Object {
    "depth": 0,
    "href": "sl/org/project/docs/getting-started.md",
    "icon": "document",
    "id": 1,
    "name": "Getting Started",
    "type": "item",
  },
  Object {
    "depth": 0,
    "href": "sl/org/project/docs/welcome.md",
    "icon": "document",
    "id": 0,
    "name": "Welcome",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": "folder-close",
    "id": "1-0",
    "name": "Guides",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/docs/guides/authentication.md",
    "icon": "document",
    "id": 2,
    "name": "Authentication",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": "folder-close",
    "id": "2-0",
    "name": "Tutorials",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/docs/tutorials/billing.md",
    "icon": "document",
    "id": 4,
    "name": "Billing",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/docs/tutorials/signup.md",
    "icon": "document",
    "id": 3,
    "name": "Signup",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": "chevron-right",
    "id": 11,
    "meta": "v1.0",
    "name": "Another API",
    "type": "divider",
  },
  Object {
    "depth": 0,
    "href": "sl/org/project/openapi.v1.yml",
    "icon": "document",
    "id": "11-overview",
    "name": "Overview",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": "folder-close",
    "id": "11-things-0",
    "name": "Things",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/openapi.v1.yml/paths/~1things/get",
    "icon": "document",
    "id": 12,
    "name": "List Things",
    "type": "item",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/openapi.v1.yml/definitions/thing",
    "icon": "document",
    "id": 13,
    "name": "Thing",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": "chevron-right",
    "id": 5,
    "name": "Todos API",
    "type": "divider",
  },
  Object {
    "depth": 0,
    "href": "sl/org/project/reference/todos/openapi.v1.yml",
    "icon": "document",
    "id": "5-overview",
    "name": "Overview",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": "folder-close",
    "id": "5-users-0",
    "name": "Users",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/reference/todos/openapi.v1.yml/paths/~1users/get",
    "icon": "document",
    "id": 6,
    "name": "Get User",
    "type": "item",
  },
  Object {
    "depth": 0,
    "icon": "folder-close",
    "id": "5-other",
    "name": "Other",
    "type": "group",
  },
  Object {
    "depth": 1,
    "href": "sl/org/project/reference/todos/openapi.v1.yml/paths/~1todos/get",
    "icon": "document",
    "id": 8,
    "name": "Get Todo",
    "type": "item",
  },
  Object {
    "depth": 0,
    "id": "models",
    "name": "Models",
    "type": "divider",
  },
  Object {
    "depth": 0,
    "href": "sl/org/project/models/example",
    "icon": "document",
    "id": 10,
    "meta": "v1.0",
    "name": "Example",
    "type": "item",
  },
  Object {
    "depth": 0,
    "href": "sl/org/project/reference/todos/models/todos",
    "icon": "document",
    "id": 9,
    "name": "Todo",
    "type": "item",
  },
  Object {
    "depth": 0,
    "href": "sl/org/project/reference/todos/models/user",
    "icon": "document",
    "id": 7,
    "meta": "v1.0",
    "name": "User",
    "type": "item",
  },
]
`;
